File: backend/apps/catalog/admin/resources.py
from import_export import resources, fields
from import_export.widgets import ForeignKeyWidget, ManyToManyWidget

from apps.catalog.models import (
    Country, Genre, RecordLabel, Person, Artist,
    Album, Song, SongCredit, OrchestraMember,
)


class CountryResource(resources.ModelResource):
    class Meta:
        model = Country
        import_id_fields = ['iso_code']
        fields = ('id', 'name', 'iso_code')


class GenreResource(resources.ModelResource):
    class Meta:
        model = Genre
        import_id_fields = ['slug']
        fields = ('id', 'name', 'slug', 'description')


class PersonResource(resources.ModelResource):
    nationality = fields.Field(
        column_name='nationality',
        attribute='nationality',
        widget=ForeignKeyWidget(Country, 'iso_code'),
    )

    class Meta:
        model = Person
        import_id_fields = ['slug']
        fields = ('id', 'full_name', 'slug', 'birth_date', 'death_date', 'nationality', 'bio')


class ArtistResource(resources.ModelResource):
    nationality = fields.Field(
        column_name='nationality',
        attribute='nationality',
        widget=ForeignKeyWidget(Country, 'iso_code'),
    )
    genres = fields.Field(
        column_name='genres',
        attribute='genres',
        widget=ManyToManyWidget(Genre, field='slug', separator='|'),
    )

    class Meta:
        model = Artist
        import_id_fields = ['slug']
        fields = ('id', 'name', 'slug', 'artist_type', 'bio', 'formed_year', 'disbanded_year', 'nationality', 'genres')


class AlbumResource(resources.ModelResource):
    artist = fields.Field(
        column_name='artist',
        attribute='artist',
        widget=ForeignKeyWidget(Artist, 'slug'),
    )
    record_label = fields.Field(
        column_name='record_label',
        attribute='record_label',
        widget=ForeignKeyWidget(RecordLabel, 'slug'),
    )
    genres = fields.Field(
        column_name='genres',
        attribute='genres',
        widget=ManyToManyWidget(Genre, field='slug', separator='|'),
    )

    class Meta:
        model = Album
        import_id_fields = ['slug']
        fields = ('id', 'title', 'slug', 'artist', 'release_year', 'release_date', 'record_label', 'album_type', 'genres')


class SongResource(resources.ModelResource):
    genres = fields.Field(
        column_name='genres',
        attribute='genres',
        widget=ManyToManyWidget(Genre, field='slug', separator='|'),
    )

    class Meta:
        model = Song
        import_id_fields = ['slug']
        fields = ('id', 'title', 'slug', 'year', 'duration', 'lyrics', 'genres')


class SongCreditResource(resources.ModelResource):
    song = fields.Field(
        column_name='song',
        attribute='song',
        widget=ForeignKeyWidget(Song, 'slug'),
    )
    person = fields.Field(
        column_name='person',
        attribute='person',
        widget=ForeignKeyWidget(Person, 'slug'),
    )

    class Meta:
        model = SongCredit
        import_id_fields = ['id']
        fields = ('id', 'song', 'person', 'role', 'instrument', 'as_name')


class OrchestraMemberResource(resources.ModelResource):
    artist = fields.Field(
        column_name='artist',
        attribute='artist',
        widget=ForeignKeyWidget(Artist, 'slug'),
    )
    person = fields.Field(
        column_name='person',
        attribute='person',
        widget=ForeignKeyWidget(Person, 'slug'),
    )

    class Meta:
        model = OrchestraMember
        import_id_fields = ['id']
        fields = ('id', 'artist', 'person', 'role', 'instrument', 'join_year', 'leave_year', 'is_current')

File: frontend/src/types/resources.ts
export interface Country {
  id: number
  name: string
  iso_code: string
}

export interface Genre {
  id: number
  name: string
  slug: string
  description: string
}

export interface RecordLabel {
  id: number
  name: string
  slug: string
  country: number | null
  country_name: string | null
  founded_year: number | null
}

export interface Person {
  id: number
  full_name: string
  slug: string
  birth_date: string | null
  death_date: string | null
  nationality: number | null
  nationality_name: string | null
  bio: string
}

export interface Artist {
  id: number
  name: string
  slug: string
  artist_type: 'solo' | 'orchestra'
  bio: string
  formed_year: number | null
  disbanded_year: number | null
  nationality: number | null
  nationality_name: string | null
  genres: number[]
}

export interface Album {
  id: number
  title: string
  slug: string
  artist: number
  artist_name: string | null
  release_year: number | null
  release_date: string | null
  record_label: number | null
  record_label_name: string | null
  album_type: 'studio' | 'live' | 'compilation' | 'ep' | 'single'
  genres: number[]
}

export interface Song {
  id: number
  title: string
  slug: string
  duration: number | null
  year: number | null
  lyrics: string
  genres: number[]
}

export interface Track {
  id: number
  album: number
  song: number
  song_title: string | null
  track_number: number
  disc_number: number
  duration_override: number | null
  is_bonus: boolean
}

export interface SongCredit {
  id: number
  song: number
  song_title: string | null
  person: number
  person_name: string | null
  role: string
  role_display: string | null
  instrument: string
  as_name: string
}
